reports/gitlab-commit-protocol/gitlab-commit-protocol.tex

```diff
@@ -98,7 +98,34 @@ \section{3PC}
 
 \section{Recovery protocol}
 
-TODO: how the GitLab handles node failures both RMs and TMs.
+
+\emph{\href{https://gitlab.com/gitlab-org/gitaly/-/blob/master/internal/praefect/nodes/sql_elector.go}{sqlElector }}manages the primary election for one virtual storage (aka
+shard). It enables multiple, redundant Praefect processes to run,
+which is needed to eliminate a single point of failure in Gitaly High
+Avaiability.
+
+The sqlElector is responsible for:
+\begin{itemize}
+\item Monitoring and updating the status of all nodes within the shard.
+\item Electing a new primary of the shard based on the health.
+\end{itemize}
+Also very Praefect node periodically performs a health check RPC with a Gitaly node.  Once the health checks are complete, Praefect node does a \textit{SELECT} from \textit{node\_status} table to determine healthy nodes.
+
+The primary of each shard is listed in the
+\textit{shard\_primaries}. If the current primary is in the healthy
+node list, then \textit{sqlElector} updates its internal state to match.
```

Comment:
What kind of list - ordered by some sign or not? How system choose which node should be taken from it?